hal/src/common/thumbv7em/clock/v2/sources/dfll.rs

```diff
@@ -1,103 +1,344 @@
+    __: (),
+}
 
 impl Registers {
     /// TODO
     #[inline]
     unsafe fn new() -> Self {
-        Self
+        Self { __: () }
     }
 
-    #[allow(dead_code)]
     #[inline]
-    fn oscctrl(&self) -> &RegisterBlock {
+    fn oscctrl(&self) -> &crate::pac::oscctrl::RegisterBlock {
         unsafe { &*crate::pac::OSCCTRL::ptr() }
     }
+
+    #[allow(dead_code)]
+    #[inline]
+    fn dfllctrla(&self) -> &crate::pac::oscctrl::DFLLCTRLA {
+        &self.oscctrl().dfllctrla
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn dfllctrlb(&self) -> &crate::pac::oscctrl::DFLLCTRLB {
+        &self.oscctrl().dfllctrlb
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn dfllval(&self) -> &crate::pac::oscctrl::DFLLVAL {
+        &self.oscctrl().dfllval
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn dfllmul(&self) -> &crate::pac::oscctrl::DFLLMUL {
+        &self.oscctrl().dfllmul
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn dfllsync(&self) -> &crate::pac::oscctrl::DFLLSYNC {
+        &self.oscctrl().dfllsync
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn wait_sync_enable(&self) {
+        while self.dfllsync().read().enable().bit() == true {}
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn wait_sync_dfllmul(&self) {
+        while self.dfllsync().read().dfllmul().bit() == true {}
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn wait_sync_dfllval(&self) {
+        while self.dfllsync().read().dfllval().bit() == true {}
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn wait_sync_dfllctrlb(&self) {
+        while self.dfllsync().read().dfllctrlb().bit() == true {}
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn enable(&mut self) {
+        self.dfllctrla().modify(|_, w| w.enable().set_bit());
+        self.wait_sync_enable();
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn set_open_mode(&mut self) {
+        self.dfllctrlb().modify(|_, w| w.mode().clear_bit());
+        self.wait_sync_enable();
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn set_closed_mode(&mut self) {
+        self.dfllctrlb().modify(|_, w| w.mode().set_bit());
+        self.wait_sync_enable();
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn set_fine_maximum_step(&mut self, value: u8) {
+        self.dfllmul()
+            .modify(|_, w| unsafe { w.fstep().bits(value) });
+        self.wait_sync_dfllmul();
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn set_coarse_maximum_step(&mut self, value: u8) {
+        self.dfllmul()
+            .modify(|_, w| unsafe { w.cstep().bits(value) });
+        self.wait_sync_dfllmul();
+    }
+    #[allow(dead_code)]
+    #[inline]
+    fn set_multiplication_factor(&mut self, value: u16) {
+        self.dfllmul().modify(|_, w| unsafe { w.mul().bits(value) });
+        self.wait_sync_dfllmul();
+    }
+}
+
+type MultiplicationFactor = u16;
+type CoarseMaximumStep = u8;
+type FineMaximumStep = u8;
+type Fine = u8;
+type Coarse = u8;
+
+pub trait Mode: Sealed {}
```

Comment:
I would suggest `LoopMode` here, and I would rename `OpenMode` and `ClosedMode` to `OpenLoop` and `ClosedLoop`. I think those are more descriptive and clear.

Reply:
Done.